File: tests/youtubelivedetector.py
#!/usr/bin/env python3
import json
import re
import sys
import urllib.request

IS_LIVE_RE = re.compile(r'"isLiveNow":(true|false)')
VIDEO_ID_IN_STRING_RE = re.compile(r'(?:/vi/|[?&]v=)([A-Za-z0-9_-]{11})')
YT_INITIAL_DATA_RE = re.compile(r"var ytInitialData\s*=\s*(\{.*?\});", re.S)


def fetch(url: str, verbose: bool) -> str:
    if verbose:
        print(f"[fetch] GET {url}")
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=20) as resp:
        data = resp.read().decode('utf-8', errors='ignore')
        if verbose:
            print(f"[fetch] status={resp.status} bytes={len(data)}")
        return data


def extract_video_id_from_url(url: str) -> str | None:
    idx = url.lower().find("watch?v=")
    if idx < 0:
        return None
    part = url[idx + len("watch?v=") :]
    amp = part.find('&')
    if amp >= 0:
        part = part[:amp]
    return part if len(part) == 11 else None


def normalize_channel_url(url: str) -> str:
    trimmed = url.rstrip('/')
    lowered = trimmed.lower()
    for suffix in ("/streams", "/live", "/videos", "/featured"):
        if lowered.endswith(suffix):
            return trimmed[: -len(suffix)]
    return trimmed


def build_streams_url(url: str) -> str:
    base = normalize_channel_url(url)
    return base + "/streams"


def is_video_id(value: str) -> bool:
    return len(value) == 11 and all(
        "A" <= c <= "Z" or "a" <= c <= "z" or "0" <= c <= "9" or c in "_-"
        for c in value
    )


def extract_video_id_from_string(value: str | None) -> str | None:
    if not value:
        return None
    if is_video_id(value):
        return value
    match = VIDEO_ID_IN_STRING_RE.search(value)
    return match.group(1) if match else None


def is_live_thumbnail_badge(obj: dict) -> bool:
    style = obj.get("badgeStyle")
    if isinstance(style, str) and style.lower() == "thumbnail_overlay_badge_style_live":
        return True

    text = obj.get("text")
    return isinstance(text, str) and "ライブ" in text


def has_live_badge(obj) -> bool:
    if isinstance(obj, dict):
        badge = obj.get("thumbnailBadgeViewModel")
        if isinstance(badge, dict) and is_live_thumbnail_badge(badge):
            return True
        return any(has_live_badge(value) for value in obj.values())
    if isinstance(obj, list):
        return any(has_live_badge(value) for value in obj)
    return False


def find_video_id_reference(obj) -> str | None:
    if isinstance(obj, dict):
        for value in obj.values():
            if isinstance(value, str):
                video_id = extract_video_id_from_string(value)
                if video_id:
                    return video_id
        for value in obj.values():
            video_id = find_video_id_reference(value)
            if video_id:
                return video_id
    elif isinstance(obj, list):
        for value in obj:
            video_id = find_video_id_reference(value)
            if video_id:
                return video_id
    elif isinstance(obj, str):
        return extract_video_id_from_string(obj)
    return None


def is_likely_video_item(obj: dict) -> bool:
    return any(
        key in obj
        for key in (
            "richItemRenderer",
            "lockupViewModel",
            "videoRenderer",
            "gridVideoRenderer",
            "compactVideoRenderer",
        )
    )


def extract_live_video_id_from_initial_data(html: str) -> str | None:
    match = YT_INITIAL_DATA_RE.search(html)
    if not match:
        return None
    try:
        data = json.loads(match.group(1))
    except json.JSONDecodeError:
        return None

    live_ids: list[str] = []

    def walk(obj) -> None:
        if isinstance(obj, dict):
            video_id = obj.get("videoId")
            overlays = obj.get("thumbnailOverlays")
            if video_id and isinstance(video_id, str) and overlays:
                for overlay in overlays:
                    renderer = overlay.get("thumbnailOverlayTimeStatusRenderer")
                    if not renderer:
                        continue
                    style = renderer.get("style")
                    text = renderer.get("text") or {}
                    label = ""
                    if isinstance(text, dict):
                        if "simpleText" in text:
                            label = text.get("simpleText") or ""
                        else:
                            runs = text.get("runs") or []
                            label = "".join(r.get("text", "") for r in runs if isinstance(r, dict))
                    if (isinstance(style, str) and style.lower() == "live") or "ライブ" in label:
                        live_ids.append(video_id)
                        break
            if is_likely_video_item(obj) and has_live_badge(obj):
                video_id = find_video_id_reference(obj)
                if video_id:
                    live_ids.append(video_id)
            for value in obj.values():
                walk(value)
        elif isinstance(obj, list):
            for value in obj:
                walk(value)

    walk(data)
    return live_ids[0] if live_ids else None


def find_live_video_id(html: str) -> str | None:
    live_id = extract_live_video_id_from_initial_data(html)
    if live_id:
        return live_id
    return None


def check_watch(video_id: str, verbose: bool) -> bool:
    html = fetch(f"https://www.youtube.com/watch?v={video_id}", verbose)
    m = IS_LIVE_RE.search(html)
    if verbose:
        print(f"[watch] isLiveNow match={m.group(1) if m else 'none'}")
    return bool(m and m.group(1).lower() == "true")


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: youtubelivedetector.py <channel_or_watch_url> [--verbose]")
        return 2
    url = sys.argv[1].strip()
    verbose = "--verbose" in sys.argv[2:] or "-v" in sys.argv[2:]
    try:
        if "watch?v=" in url.lower():
            video_id = extract_video_id_from_url(url)
            if verbose:
                print(f"[input] watch url videoId={video_id or 'none'}")
            if not video_id:
                print("NOT_LIVE")
                return 0
            print("LIVE" if check_watch(video_id, verbose) else "NOT_LIVE")
            return 0

        base_url = normalize_channel_url(url)
        streams_url = build_streams_url(url)
        if verbose:
            print(f"[input] channel url base={base_url}")
            print(f"[streams] url={streams_url}")
        streams_html = fetch(streams_url, verbose)
        video_id = find_live_video_id(streams_html)
        if verbose:
            print(f"[streams] live videoId={video_id or 'none'}")

        if not video_id:
            if verbose:
                print(f"[home] url={base_url}")
            home_html = fetch(base_url, verbose)
            video_id = find_live_video_id(home_html)
            if verbose:
                print(f"[home] live videoId={video_id or 'none'}")

        if not video_id:
            print("NOT_LIVE")
            return 0
        print("LIVE" if check_watch(video_id, verbose) else "NOT_LIVE")
        return 0
    except Exception as e:
        print(f"ERROR: {e}")
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: src/LiveAlert.Core/YouTubeLiveDetector.cs
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace LiveAlert.Core;

public interface ILiveDetector
{
    Task<LiveCheckResult> CheckLiveAsync(AlertConfig alert, CancellationToken cancellationToken);
}

public sealed class YouTubeLiveDetector : ILiveDetector
{
    private readonly HttpClient _httpClient;
    private static readonly Regex IsLiveNowRegex = new("\"isLiveNow\":(true|false)", RegexOptions.Compiled);
    private static readonly Regex InitialDataRegex = new(@"var ytInitialData\s*=\s*(\{.*?\});", RegexOptions.Compiled | RegexOptions.Singleline);
    private static readonly Regex VideoIdInStringRegex = new(@"(?:/vi/|[?&]v=)([A-Za-z0-9_-]{11})", RegexOptions.Compiled);

    public YouTubeLiveDetector(HttpClient httpClient)
    {
        _httpClient = httpClient;
        if (!_httpClient.DefaultRequestHeaders.UserAgent.Any())
        {
            _httpClient.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("Mozilla", "5.0"));
        }
    }

    public async Task<LiveCheckResult> CheckLiveAsync(AlertConfig alert, CancellationToken cancellationToken)
    {
        try
        {
            var url = alert.Url.Trim();
            if (url.Contains("watch?v=", StringComparison.OrdinalIgnoreCase))
            {
                var videoId = ExtractVideoIdFromUrl(url);
                if (string.IsNullOrEmpty(videoId))
                {
                    return LiveCheckResult.NotLive();
                }

                return await CheckWatchPageAsync(videoId, cancellationToken).ConfigureAwait(false);
            }

            var baseUrl = NormalizeChannelUrl(url);
            var streamsUrl = $"{baseUrl}/streams";
            var (streamsOk, streamsHtml, streamsError) = await FetchStringAsync(streamsUrl, cancellationToken).ConfigureAwait(false);
            if (!streamsOk)
            {
                return streamsError ?? LiveCheckResult.Error("HTTP error");
            }

            var candidateId = ExtractLiveVideoId(streamsHtml);
            if (string.IsNullOrEmpty(candidateId))
            {
                var (homeOk, homeHtml, homeError) = await FetchStringAsync(baseUrl, cancellationToken).ConfigureAwait(false);
                if (!homeOk)
                {
                    return homeError ?? LiveCheckResult.Error("HTTP error");
                }
                candidateId = ExtractLiveVideoId(homeHtml);
            }

            if (string.IsNullOrEmpty(candidateId))
            {
                return LiveCheckResult.NotLive();
            }

            return await CheckWatchPageAsync(candidateId, cancellationToken).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw;
        }
        catch (OperationCanceledException ex)
        {
            return LiveCheckResult.Error(ex.Message);
        }
        catch (Exception ex)
        {
            return LiveCheckResult.Error(ex.Message);
        }
    }

    private async Task<LiveCheckResult> CheckWatchPageAsync(string videoId, CancellationToken cancellationToken)
    {
        var watchUrl = $"https://www.youtube.com/watch?v={videoId}";
        var (watchOk, html, watchError) = await FetchStringAsync(watchUrl, cancellationToken).ConfigureAwait(false);
        if (!watchOk)
        {
            return watchError ?? LiveCheckResult.Error("HTTP error");
        }
        var liveMatch = IsLiveNowRegex.Match(html ?? string.Empty);
        if (liveMatch.Success && liveMatch.Groups[1].Value.Equals("true", StringComparison.OrdinalIgnoreCase))
        {
            return LiveCheckResult.Live(videoId);
        }

        return LiveCheckResult.NotLive();
    }

    private static string NormalizeChannelUrl(string url)
    {
        url = url.TrimEnd('/');
        var lowered = url.ToLowerInvariant();
        foreach (var suffix in new[] { "/streams", "/live", "/videos", "/featured" })
        {
            if (lowered.EndsWith(suffix, StringComparison.Ordinal))
            {
                return url[..^suffix.Length];
            }
        }
        return url;
    }

    private static string? ExtractLiveVideoId(string? html)
    {
        if (string.IsNullOrEmpty(html))
        {
            return null;
        }

        var match = InitialDataRegex.Match(html);
        if (!match.Success)
        {
            return null;
        }

        try
        {
            using var document = JsonDocument.Parse(match.Groups[1].Value);
            return FindLiveVideoId(document.RootElement);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    private static string? ExtractVideoIdFromUrl(string url)
    {
        var idx = url.IndexOf("watch?v=", StringComparison.OrdinalIgnoreCase);
        if (idx < 0) return null;
        var part = url[(idx + "watch?v=".Length)..];
        var amp = part.IndexOf('&');
        if (amp >= 0)
        {
            part = part[..amp];
        }
        return part.Length == 11 ? part : null;
    }

    private async Task<(bool ok, string? content, LiveCheckResult? error)> FetchStringAsync(string url, CancellationToken cancellationToken)
    {
        using var response = await _httpClient.GetAsync(url, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            return (false, null, LiveCheckResult.Error($"HTTP {(int)response.StatusCode} {response.ReasonPhrase}"));
        }

        var html = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        return (true, html, null);
    }

    private static string? FindLiveVideoId(JsonElement element)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.Object:
            {
                if (element.TryGetProperty("videoId", out var videoIdElement) &&
                    videoIdElement.ValueKind == JsonValueKind.String &&
                    element.TryGetProperty("thumbnailOverlays", out var overlaysElement) &&
                    overlaysElement.ValueKind == JsonValueKind.Array)
                {
                    if (HasLiveOverlay(overlaysElement))
                    {
                        return videoIdElement.GetString();
                    }
                }

                if (IsLikelyVideoItem(element) &&
                    HasLiveBadge(element) &&
                    TryFindVideoIdReference(element, out var liveItemVideoId))
                {
                    return liveItemVideoId;
                }

                foreach (var property in element.EnumerateObject())
                {
                    var found = FindLiveVideoId(property.Value);
                    if (!string.IsNullOrEmpty(found))
                    {
                        return found;
                    }
                }

                break;
            }
            case JsonValueKind.Array:
            {
                foreach (var item in element.EnumerateArray())
                {
                    var found = FindLiveVideoId(item);
                    if (!string.IsNullOrEmpty(found))
                    {
                        return found;
                    }
                }

                break;
            }
        }

        return null;
    }

    private static bool IsLikelyVideoItem(JsonElement element)
    {
        return element.TryGetProperty("richItemRenderer", out _) ||
               element.TryGetProperty("lockupViewModel", out _) ||
               element.TryGetProperty("videoRenderer", out _) ||
               element.TryGetProperty("gridVideoRenderer", out _) ||
               element.TryGetProperty("compactVideoRenderer", out _);
    }

    private static bool HasLiveBadge(JsonElement element)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.Object:
            {
                if (element.TryGetProperty("thumbnailBadgeViewModel", out var badge) &&
                    badge.ValueKind == JsonValueKind.Object &&
                    IsLiveThumbnailBadge(badge))
                {
                    return true;
                }

                foreach (var property in element.EnumerateObject())
                {
                    if (HasLiveBadge(property.Value))
                    {
                        return true;
                    }
                }

                break;
            }
            case JsonValueKind.Array:
            {
                foreach (var item in element.EnumerateArray())
                {
                    if (HasLiveBadge(item))
                    {
                        return true;
                    }
                }

                break;
            }
        }

        return false;
    }

    private static bool IsLiveThumbnailBadge(JsonElement badge)
    {
        if (badge.TryGetProperty("badgeStyle", out var badgeStyle) &&
            badgeStyle.ValueKind == JsonValueKind.String &&
            badgeStyle.GetString()?.Equals("THUMBNAIL_OVERLAY_BADGE_STYLE_LIVE", StringComparison.OrdinalIgnoreCase) == true)
        {
            return true;
        }

        if (badge.TryGetProperty("text", out var text) &&
            text.ValueKind == JsonValueKind.String &&
            text.GetString()?.Contains("ライブ", StringComparison.OrdinalIgnoreCase) == true)
        {
            return true;
        }

        return false;
    }

    private static bool TryFindVideoIdReference(JsonElement element, out string videoId)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.Object:
            {
                foreach (var property in element.EnumerateObject())
                {
                    if (property.Value.ValueKind == JsonValueKind.String &&
                        TryExtractVideoIdFromString(property.Value.GetString(), out videoId))
                    {
                        return true;
                    }
                }

                foreach (var property in element.EnumerateObject())
                {
                    if (TryFindVideoIdReference(property.Value, out videoId))
                    {
                        return true;
                    }
                }

                break;
            }
            case JsonValueKind.Array:
            {
                foreach (var item in element.EnumerateArray())
                {
                    if (TryFindVideoIdReference(item, out videoId))
                    {
                        return true;
                    }
                }

                break;
            }
            case JsonValueKind.String:
                if (TryExtractVideoIdFromString(element.GetString(), out videoId))
                {
                    return true;
                }
                break;
        }

        videoId = string.Empty;
        return false;
    }

    private static bool TryExtractVideoIdFromString(string? value, out string videoId)
    {
        if (!string.IsNullOrEmpty(value))
        {
            if (value.Length == 11 && value.All(IsVideoIdChar))
            {
                videoId = value;
                return true;
            }

            var match = VideoIdInStringRegex.Match(value);
            if (match.Success)
            {
                videoId = match.Groups[1].Value;
                return true;
            }
        }

        videoId = string.Empty;
        return false;
    }

    private static bool IsVideoIdChar(char value)
    {
        return value is >= 'A' and <= 'Z' ||
               value is >= 'a' and <= 'z' ||
               value is >= '0' and <= '9' ||
               value == '_' ||
               value == '-';
    }

    private static bool HasLiveOverlay(JsonElement overlaysElement)
    {
        foreach (var overlay in overlaysElement.EnumerateArray())
        {
            if (!overlay.TryGetProperty("thumbnailOverlayTimeStatusRenderer", out var renderer) ||
                renderer.ValueKind != JsonValueKind.Object)
            {
                continue;
            }

            if (renderer.TryGetProperty("style", out var styleElement) &&
                styleElement.ValueKind == JsonValueKind.String &&
                styleElement.GetString()?.Equals("LIVE", StringComparison.OrdinalIgnoreCase) == true)
            {
                return true;
            }

            if (renderer.TryGetProperty("text", out var textElement) &&
                TextContainsLiveLabel(textElement))
            {
                return true;
            }
        }

        return false;
    }

    private static bool TextContainsLiveLabel(JsonElement textElement)
    {
        if (textElement.ValueKind != JsonValueKind.Object)
        {
            return false;
        }

        if (textElement.TryGetProperty("simpleText", out var simpleText) &&
            simpleText.ValueKind == JsonValueKind.String &&
            simpleText.GetString()?.Contains("ライブ", StringComparison.OrdinalIgnoreCase) == true)
        {
            return true;
        }

        if (textElement.TryGetProperty("runs", out var runsElement) &&
            runsElement.ValueKind == JsonValueKind.Array)
        {
            foreach (var run in runsElement.EnumerateArray())
            {
                if (run.ValueKind != JsonValueKind.Object)
                {
                    continue;
                }

                if (run.TryGetProperty("text", out var runText) &&
                    runText.ValueKind == JsonValueKind.String &&
                    runText.GetString()?.Contains("ライブ", StringComparison.OrdinalIgnoreCase) == true)
                {
                    return true;
                }
            }
        }

        return false;
    }
}

public readonly record struct LiveCheckResult(bool IsLive, string? VideoId, bool IsError, string? ErrorMessage)
{
    public static LiveCheckResult Live(string videoId) => new(true, videoId, false, null);
    public static LiveCheckResult NotLive() => new(false, null, false, null);
    public static LiveCheckResult Error(string message) => new(false, null, true, message);
}
